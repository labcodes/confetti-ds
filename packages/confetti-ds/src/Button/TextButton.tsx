import React from "react";

import AbstractButton from "./AbstractButton";
import { BaseButtonProps } from "./types";

export interface TextButtonProps extends BaseButtonProps {
  theme?: "teal" | "purple" | "light" | "dark";
}

export default function TextButton({
  label,
  type = "button",
  theme = "teal",
  hasIcon = false,
  icon,
  size = "normal",
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  fullWidth = false,
  tabIndex,
}: TextButtonProps) {
  return (
    <AbstractButton
      kind="text"
      label={label}
      type={type}
      theme={theme}
      icon={icon}
      hasIcon={hasIcon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}
