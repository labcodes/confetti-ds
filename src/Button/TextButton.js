import React from "react";
import PropTypes from "prop-types";

import AbstractButton from "./AbstractButton";
import { ICON_TYPES } from "../constants";

export default class TextButton extends React.Component {
  static propTypes = {
    /** This is the button label. */
    text: PropTypes.string.isRequired,
    /** HTML type of the Button. */
    type: PropTypes.oneOf(["button", "submit", "reset"]),
    /** Sets a special color skin to the button. */
    skin: PropTypes.oneOf(["", "light", "dark"]),
    /** Sets the icon related to the button label. Default state: no icon. */
    icon: PropTypes.oneOf(ICON_TYPES),
    /** Sets the button's height. Small = 32px, Normal = 40px, Large = 48px. */
    size: PropTypes.oneOf(["normal", "small", "large"]),
    /** Disables the Button. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Button. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Action to be executed when the button is clicked. */
    onClick: PropTypes.func,
    /** Makes the button expand to its container's full width. */
    fullWidth: PropTypes.bool,
  };

  static defaultProps = {
    skin: undefined,
    icon: undefined,
    type: "button",
    size: "normal",
    disabled: false,
    ariaDisabled: false,
    onClick: () => {},
    fullWidth: false,
  };

  render() {
    const {
      text,
      type,
      skin,
      icon,
      size,
      disabled,
      onClick,
      fullWidth,
      ariaDisabled,
    } = this.props;
    return (
      <AbstractButton
        variant="text"
        text={text}
        type={type}
        skin={skin}
        icon={icon}
        size={size}
        disabled={disabled}
        ariaDisabled={ariaDisabled}
        onClick={onClick}
        fullWidth={fullWidth}
      />
    );
  }
}
