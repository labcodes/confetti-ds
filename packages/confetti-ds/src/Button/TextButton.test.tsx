import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import TextButton from "./TextButton";

describe("TextButton", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text Disabled Button 1" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TextButton label="Test Outline Disabled Button 2" disabled />
    );
    expect(mountedComponent.find("button").prop("disabled")).toEqual(true);
  });

  it("renders as expected when passing a dark skin", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text Dark Button" theme="dark" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a small size", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text Small Button" size="small" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a plus icon", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text Small Button" icon="plus" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("calls props.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowButton = shallow(
      <TextButton label="Test Click on Text Button" onClick={mockOnClick} />
    );
    shallowButton.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it("doesn't call onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const mountedComponent = mount(
      <TextButton ariaDisabled onClick={mockOnClick} label="Test Text Button" />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    mountedComponent.find("button").simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(0);
  });

  it("renders as expected if full width", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test Text fullWidth Button" fullWidth />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with full width", async () => {
    const mountedComponent = mount(
      <TextButton label="Test fullWidth Button" fullWidth />
    );
    expect(mountedComponent.find(".lab-btn--block")).toHaveLength(1);
  });

  it("renders as expected when tabIndex is passed", async () => {
    const renderedComponent = renderer
      .create(<TextButton label="Test tabIndex" tabIndex={1} />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
