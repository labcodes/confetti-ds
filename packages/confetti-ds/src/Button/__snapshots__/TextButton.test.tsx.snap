// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextButton renders as expected if full width 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal lab-btn--block"
  onClick={[Function]}
  type="button"
>
  
  Test Text fullWidth Button
</button>
`;

exports[`TextButton renders as expected when passing a dark skin 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal lab-btn--dark"
  onClick={[Function]}
  type="button"
>
  
  Test Text Dark Button
</button>
`;

exports[`TextButton renders as expected when passing a plus icon 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  <span
    className="lab-icon lab-icon--Add lab-icon--white lab-btn__icon"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height={20}
      viewBox="0 0 20 20"
      width={20}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 2.5a.5.5 0 0 0-1 0V9H2.5a.5.5 0 0 0 0 1H9v6.5a.5.5 0 0 0 1 0V10h6.5a.5.5 0 0 0 0-1H10V2.5Z"
        fill="currentColor"
      />
    </svg>
  </span>
  Test Text Small Button
</button>
`;

exports[`TextButton renders as expected when passing a small size 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--small"
  onClick={[Function]}
  type="button"
>
  
  Test Text Small Button
</button>
`;

exports[`TextButton renders as expected when passing disabled as true 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  
  Test Text Disabled Button 1
</button>
`;

exports[`TextButton renders as expected when tabIndex is passed 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal"
  onClick={[Function]}
  tabIndex={1}
  type="button"
>
  
  Test tabIndex
</button>
`;

exports[`TextButton renders with base props 1`] = `
<button
  className="lab-btn lab-btn--text lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Text Button
</button>
`;
